### PR TITLE
Use courier for logging/error handling in hpwh

### PR DESCRIPTION
**Pull request by spahrenk**
## Description

Logging and error handling in HPWH changed to use Courier. Calls to HPWH functions are now enclosed in try/catch statements. A shared pointer to an object of the Courier-derived type `CourierMsgHandler` is constructed prior to instantiating a HPWH object and passed to the constructor. The `MessageLevel` enum class was duplicated from the Courier examples, with a corresponding member value included in `CourierMsgHandlerBase`.

### CODE REVIEW

**Review by nealkruis — CHANGES_REQUESTED** [review]
I'm reviewing all the branches that are part of [`develop-hpwh`](https://github.com/cse-sim/cse/tree/develop-hpwh). This branch needs to remove the "try/catches" everywhere. It is not possible for these functions to throw because the callback will ensure termination before the exception gets to this level.

**Comment by spahrenk** [comment]
Removed try/catches.

File: src/dhwcalc.cpp [hunk 1]
```diff
@@ -3400,35 +3393,35 @@ RC HPWHLINK::hw_DeriveVolFromVolRunning(		// calc required volume from running v
...
-		useableFract = .9f;
-	}
-	useableFract = bracket(.6, useableFract, 1.);
-	double unuseableFract = 1. - useableFract;
-	aquaFract = bracket(unuseableFract + .1, aquaFract, .75);
 
-	// total volume req'd based on aquastat position
-	//    Running vol is vol above aquastat
-	float totVolRun = float(volRunning / (1. - aquaFract));
+    if (hw_pHPWH->getSizingFractions(aquaFract, useableFract) != 0) {
```
Comment (nealkruis) [inline src/dhwcalc.cpp:3401]: It looks like this function always returns zero. It should be modified to either return the two fractions or be a void function.

File: src/dhwcalc.cpp [hunk 2]
```diff
@@ -3477,18 +3470,28 @@ double HPWHLINK::hw_GetTankAvgTemp(		// average temp of range of tank nodes
 	int incr = nNodes < 0 ? -1 : 1;
 
 	double T = 0.;
-	for (int iN=iNode0; iN != iNodeN; iN+=incr)
-		T += hw_pHPWH->getTankNodeTemp(iN, HPWH::UNITS_C);
-	T /= max(1, abs(iNodeN - iNode0));
+	try {
```
Comment (nealkruis) [inline src/dhwcalc.cpp:3473]: Remove try/catch.

File: src/dhwcalc.cpp [hunk 3]
```diff
@@ -3477,18 +3470,28 @@ double HPWHLINK::hw_GetTankAvgTemp(		// average temp of range of tank nodes
...
 }		// HPWHLINK::hw_GetTankAvgTemp
 //-----------------------------------------------------------------------------
 double HPWHLINK::hw_GetEstimatedTOut() const
 // returns estimate of tank output temp, F
 //   = current top node temp (no consideration of draw etc.)
 {
-	int iNodeTop = hw_pHPWH->getNumNodes() - 1;
-	return hw_pHPWH->getTankNodeTemp(iNodeTop, HPWH::UNITS_F);
+	double T = 0.;
+	try {
```
Comment (nealkruis) [inline src/dhwcalc.cpp:3488]: Remove try/catch.

File: src/dhwcalc.cpp [hunk 4]
```diff
@@ -3515,57 +3518,49 @@ RC HPWHLINK::hw_DoHour(		// hourly HPWH calcs
...
+    }
+
+    // retrieve resulting setpoint after HPWH restrictions
+    tSetpoint = hw_pHPWH->getSetpoint(HPWH::UNITS_F);
+    if (hw_tHWOut == 0.f)
+        hw_tHWOut = tSetpoint; // initial guess for HW output temp
+                                //   updated every substep with nz draw
+
+    // tank temp initialization
+    if (!hw_tankTempSet) { // initialize tank temp on 1st call
```
Comment (nealkruis) [inline src/dhwcalc.cpp:3544]: Accidental formatting of comment? Should be on next line aligned with other comments.

File: src/dhwcalc.cpp [hunk 5]
```diff
@@ -4706,16 +4700,18 @@ RC DHWHEATER::wh_HPWHInit()		// initialize HPWH model
...
-			int ret = wh_HPWH.hw_pHPWH->switchToSoCControls(
-				0.9,	// initial target SoC (altered hourly, see DHWHEATER::wh_DoHour())
-				0.05,	// hysteresis
-				tMin,	// reference temp (= min useful temp)
-				false,	// mains temp varies
-				55.,	// placeholder mains temp
-				HPWH::UNITS_F);	// temps are F
-
-			if (ret != 0)
+			try {
```
Comment (nealkruis) [inline src/dhwcalc.cpp:4703]: Remove try/catch.

**Comment by spahrenk** [comment]
Should be ready for review.

File: src/dhwcalc.cpp [hunk 6]
```diff
@@ -3383,11 +3370,17 @@ float HPWHLINK::hw_GetTankSurfaceArea(		// tank surface area
 // returns total tank surface area, ft2
 //   (accounting for possible multiple tanks)
 {
-	if (vol < 0)
-		vol = hw_pHPWH->getTankSize(HPWH::UNITS_GAL);
-	float volPerTank = vol / tankCount;
-	double surfAPerTank = HPWH::getTankSurfaceArea(volPerTank, HPWH::UNITS_GAL, HPWH::UNITS_FT2);
-	float surfA = surfAPerTank * tankCount;
+	float surfA = 0.;
+	try {
```
Comment (nealkruis) [inline src/dhwcalc.cpp:3374]: One more...

File: src/dhwcalc.cpp [hunk 7]
```diff
@@ -24,6 +24,7 @@
 
 #include "cnguts.h"
 #include "exman.h"
+#include "hvac.h"
```
Comment (nealkruis) [inline src/dhwcalc.cpp:27]: Why is this needed? Please add comment to indicate its use.

**Review by nealkruis — CHANGES_REQUESTED** [review]

File: src/dhwcalc.cpp (hunk 7 above)
Comment (spahrenk) [reply src/dhwcalc.cpp:27]: Existing definitions for derived courier class were in "hvac.h". (Related data structures are mentioned in the PR description.)

Comment (nealkruis) [reply src/dhwcalc.cpp:27]: Please add this comment to the file (not just on GitHub). i.e.:

```c++
#include "hvac.h" // CourierMsgHandler, CourierMsgHandlerBase, MessageLevel
```

Comment (spahrenk) [reply src/dhwcalc.cpp]: Done.

File: src/dhwcalc.cpp (hunk 6 above)
Comment (nealkruis) [reply src/dhwcalc.cpp:3374]: @spahrenk this also still needs to be addressed.
